### PR TITLE
5 incorporate normalization factors

### PR DESCRIPTION
**Pull request by arthur-lin1027**
Starting a PR so I can show what each normalization method does.

### CODE REVIEW

**Comment by arthur-lin1027** [comment]
Commit [3ad3c6c](https://github.com/cersonsky-lab/anisoap/pull/6/commits/3ad3c6c6fccab2d3033bd5f749c87a1290aca3de) normalizes post-hoc and looks nice, but likely isn't correct because it only contains a normalization factor consisting only of the gto's gaussian width, not of the 3 axes of our anisotropic gaussian.

This results in a anisoap vector (at the end of doing the tensor product) that looks like this:

image


To obtain a normalization factor consisting of all three gaussian widths, I normalized within the pairwise expansion, as seen in this commit: [cdb48a2](https://github.com/cersonsky-lab/anisoap/pull/6/commits/cdb48a23824bbec6a3b2a9edfa1c04a94f87980d)

This results in a anisoap vector (at the end of doing the tensor product) that looks like this:

image


I'd love to hear your comments on how these look!

**Comment by arthur-lin1027** [comment]
The normalization code is ready for review. It is pretty much the same as [3ad3c6c](https://github.com/cersonsky-lab/anisoap/pull/6/commits/3ad3c6c6fccab2d3033bd5f749c87a1290aca3de)

I've included scratch code here that reads in the benzene files, and compares the normalized vs unnormalized representations. 

I'm still not sure it's quite normalizing by the correct value, but I'd like a review now because I'm kind of spinning my wheels a little bit.

**Comment by arthur-lin1027** [comment]
Am also open to feedback on what kinds of tests I should include.

I manually checked that contract(normalize(pairwise_features)) == normalize(contract(pairwise_features)), should I write that as a separate test? I see that there aren't individual tests for pairwise_ellip_expansion and contract_pairwise_feat so I wasn't sure if I should include one for the normalization function.

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (rosecers) [inline anisoap/representations/ellipsoidal_density_projection.py:432]: Do we ever want to do this without normalizing?

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (rosecers) [inline anisoap/representations/ellipsoidal_density_projection.py:520]: Better syntax to put features first, then other args after

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (rosecers) [inline anisoap/representations/ellipsoidal_density_projection.py:524]: Lol so if we don't want to normalize, we compute it anyways? See comment above about optional normalization, but if normalize=False, we shouldn't compute the normalization

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (rosecers) [inline anisoap/representations/ellipsoidal_density_projection.py:345]: I would make sure to explicitly import gamma from `scipy.special` so we have it explicitly in here and not just inherited through moment generator

File: anisoap/scratch.py
Comment (rosecers) [inline anisoap/scratch.py:103]: 👌🏻

**Comment by arthur-lin1027** [comment]
I've completed my first iteration of orthonormalization. Orthonormalization functions are now all in the radial_basis.py function. The orthonormalized  results perform well!

* First, super important to check the math. I'm pretty confident in the math but it's good to get a second pair of eyes. I've taken a slightly different approach to librascal by calculating overlap with $n_{eff}$ and $\sigma_{eff}$ but it should be equivalent (and to me, more intuitive).
* I'm super welcome to any API changes, e.g. maybe these new methods that I added should not be part of the class since they are largely static...?
* If any docstrings are confusing then let me know as well

File: anisoap/representations/radial_basis.py
Comment (rosecers) [inline anisoap/representations/radial_basis.py:41]: Small API thing -- let's call this `overlap_matrix`, so that if we're working with non-gto, we don't have a variable named something about gto's

File: anisoap/representations/radial_basis.py
Comment (rosecers) [inline anisoap/representations/radial_basis.py:127]: please clean before merging

File: anisoap/representations/radial_basis.py
Comment (rosecers) [inline anisoap/representations/radial_basis.py:101]: These would probably make more sense in util, rather than in radial_basis.

Comment (rosecers) [reply anisoap/representations/radial_basis.py:101]: Or at the very least outside of the class.

**Review by rosecers — COMMENTED** [review]
Hi Arthur --

This, in general, is looking very nice. I would encourage cleaning it up in a few ways:

- move utility functions outside of the RadialBasis class
- remove scratch.py in favor of a jupyter notebook that demonstrates the orthonormalization (no need to go all the way to anisoap for this)
- remove errant comments

Once you do this, we're near merge!

File: anisoap/representations/radial_basis.py
Comment (arthur-lin1027) [reply anisoap/representations/radial_basis.py:101]: I placed these functions outside the class

File: anisoap/representations/radial_basis.py
Comment (rosecers) [inline anisoap/representations/radial_basis.py:29]: Any reason for not using cholesky? 

https://stackoverflow.com/questions/71232192/efficient-matrix-square-root-of-large-symmetric-positive-semidefinite-matrix-in

File: anisoap/representations/radial_basis.py
Comment (rosecers) [inline anisoap/representations/radial_basis.py:28]: I'm not sure these checks belong here -- do we ever worry about them being triggered?

File: anisoap/__init__.py
Comment (rosecers) [inline anisoap/__init__.py:4]: isort is v petty

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (rosecers) [reply anisoap/representations/ellipsoidal_density_projection.py:402]: ^^

File: notebooks/GTO Orthogonalization Demonstration.ipynb
Comment (rosecers) [inline notebooks/GTO Orthogonalization Demonstration.ipynb:1]: Love this addition! We can merge as a jupyter notebook, but let's incorporate it into docs as soon as they're up.

**Review by rosecers — COMMENTED** [review]
Looking good! @curiosity54 can you take a look before we merge?

File: anisoap/representations/radial_basis.py
Comment (arthur-lin1027) [reply anisoap/representations/radial_basis.py:28]: Hmm in its current usage these checks will not be triggered. However, these are general purpose util functions that could theoretically be (mis)used elsewhere. 

So, should it be the function's role to verify that the matrix is symmetric positive semidefinite, or should such a verification be done outside the function?

**Comment by rosecers** [comment]
I could go either way — if you keep the checks, they should have coinciding tests.

File: anisoap/representations/radial_basis.py
Comment (arthur-lin1027) [reply anisoap/representations/radial_basis.py:29]: I don't think cholesky decomp == square root, since cholesky decomp is $A=LL^T$, not $A=LL$. 

This is how inverse square root was implemented by librascal and rascaline so I just used their implementation, and . If we are getting hung up on speed on this step in the future we can possibly explore a faster option.

**Comment by arthur-lin1027** [comment]
I don't know how to test for orthonormality... the numeric integration isn't great LOL. But I've been able to test for everything else.

**Comment by arthur-lin1027** [comment]
Unfortunately this code does not pass when we are dealing with large number of lmax (I'm working with lmax=15, which is what is needed for the gay berne ellipsoids). For whatever reason, our gram matrix is no longer positive semidefinite and has negative eigenvalues, which I find surprising -- I'm suspecting this is a numerical stability issue

**Review by rosecers — COMMENTED** [review]
Almost there! One or two outstanding comments, then we're good to go

File: anisoap/representations/radial_basis.py
Comment (arthur-lin1027) [reply anisoap/representations/radial_basis.py:87]: I agree -- this framework is general enough to handle different radial bases if we decide to implement them, in which case it would make more sense to move all this orthonormalization logic to a new utils. However, since we only have GTOs for now, I think making a new file containing this logic may be unnecessary indirection at the moment.

File: anisoap/representations/radial_basis.py
Comment (arthur-lin1027) [reply anisoap/representations/radial_basis.py:98]: <3

File: tests/test_ellipsoidal_density_projection.py
Comment (arthur-lin1027) [reply tests/test_ellipsoidal_density_projection.py]: OK, I added some additional tests.

File: tests/test_radial_basis.py
Comment (arthur-lin1027) [reply tests/test_radial_basis.py]: done

File: anisoap/representations/radial_basis.py
Comment (arthur-lin1027) [reply anisoap/representations/radial_basis.py:98]: I'm gonna remove that comment because I all-love-no-hate and I don't think he was actually wrong

File: notebooks/GTO Orthogonalization Demonstration.ipynb
Comment (arthur-lin1027) [reply notebooks/GTO Orthogonalization Demonstration.ipynb:1]: I'm just going to add one additional comment to the notebook saying that to ensure numerical stability in actual usage, one should orthogonalize normalized GTOs, rather than orthonormalize unnormalized GTOs.

File: anisoap/representations/radial_basis.py
Comment (Luthaf) [reply anisoap/representations/radial_basis.py]: I did not even write that code in librascal, that was Alex & friends =). There might very well be a couple of typos over there!

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (arthur-lin1027) [reply anisoap/representations/ellipsoidal_density_projection.py]: Now it's inplace.

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (arthur-lin1027) [reply anisoap/representations/ellipsoidal_density_projection.py]: Done

File: anisoap/representations/ellipsoidal_density_projection.py
Comment (arthur-lin1027) [reply anisoap/representations/ellipsoidal_density_projection.py]: Fixed this long ago

**Review by rosecers — APPROVED** [review]

**Pull request merged**
